eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Gov.hs

```diff
@@ -246,12 +258,48 @@ actionWellFormed ga = failureUnless isWellFormed $ MalformedProposal ga
       ParameterChange _ ppd -> ppuWellFormed ppd
       _ -> True
 
+checkProposalsHaveAValidPrevious ::
+  forall era.
+  PrevGovActionIds era ->
+  GovSnapshots era ->
+  GovProcedures era ->
+  Test (ConwayGovPredFailure era)
+checkProposalsHaveAValidPrevious prevGovActionIds snapshots procedures =
+  let isValidMember ::
+        StrictMaybe (PrevGovActionId p (EraCrypto era)) ->
+        (GovAction era -> Bool) ->
+        Bool
+      isValidMember prev snapshotCond =
+        case prev of
+          SNothing -> True
+          SJust prev' -> case Map.lookup (unPrevGovActionId prev') (snapshotGovActionStates $ snapshots ^. curGovSnapshotsL) of
+            Nothing -> False
+            -- lookup has to succeed _and_ purpose of looked-up action has to match condition
+            Just found -> snapshotCond . gasAction $ found
```

Comment:
This version is much clearer, however there are a few things that need to be changed:

* Whenever previous id is not provided it is invalid, unless it is the very first proposed gov action of that type. Therefore it is only valid for `SNothing` case when `prevActionAsExpected` allows it.
* `isValidMember` is a confusing name, since we have CC members and could easily be confused as related concept. Also it is not just a Map membership test, there is also `snapshotCond` application.
* `foo . bar $ baz` uses two more characters and one more function when compared to `foo $ bar baz`
* There is a specialized function for looking into snapshots: `snapshotLookupId`
```suggestion
      isValidPrevGovActionId prevGovActionId snapshotCond =
        case prevGovActionId of
          SNothing -> False
          SJust (PrevGovActionId govActionId) -> 
            case snapshotLookupId govActionId (snapshots ^. curGovSnapshotsL) of
              Nothing -> False
              -- lookup has to succeed _and_ purpose of looked-up action has to match condition
              Just found -> snapshotCond $ gasAction found
```

Reply:
I could not agree more. Thanks for the insightful catch on the `SNothing` part too! :+1: